### PR TITLE
Adds matrix function math.sqrtm(A)

### PR DESCRIPTION
**Pull request by ferrolho**
## Summary

Adds the `math.sqrtm(A)` matrix function to compute the principal square root of a square matrix. This is similar to MATLAB's [`sqrtm`](https://uk.mathworks.com/help/matlab/ref/sqrtm.html).

This PR is somewhat related to #263.

#### Denman–Beavers method

The implementation being current used is based on the [Denman–Beavers iterative method](https://en.wikipedia.org/wiki/Square_root_of_a_matrix#By_Denman%E2%80%93Beavers_iteration):

Denman–Beavers Y_{k+1}

Denman–Beavers Z_{k+1}

#### Babylonian method

The [Babylonian iterative method](https://en.wikipedia.org/wiki/Square_root_of_a_matrix#By_the_Babylonian_method) is also implemented:

Babylonian method

> *"However, unlike Denman–Beavers iteration, the Babylonian method is numerically unstable and more likely to fail to converge."* [2]

---

#### References
1. https://en.wikipedia.org/wiki/Square_root_of_a_matrix
2. Higham, N.J., 1986. [Newton’s method for the matrix square root](http://www.ams.org/journals/mcom/1986-46-174/S0025-5718-1986-0829624-5/S0025-5718-1986-0829624-5.pdf). *Mathematics of Computation*, 46(174), pp.537-549.

### CODE REVIEW

File: lib/function/matrix/sqrtm.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+   *
+   *     X = math.sqrtm(A)
+   *
+   * Examples:
+   *
+   *     math.sqrtm([[1, 2], [3, 4]]); // returns [[-2, 1], [1.5, -0.5]]
+   *
+   * See also:
+   *
+   *     det, transpose
```
Comment (harrysarson) [inline lib/function/matrix/sqrtm.js]: Are these the best suggestions? I would say replace these with more relavant functions.

File: test/function/matrix/sqrtm.test.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+         [ 1, -4,  6, -4,  1],
+         [ 0,  1, -4,  6, -4],
+         [ 0,  0,  1, -4,  6]]),
+      math.matrix(
+        [[ 2.001450032980806, -0.997069854672086,  0.004226841220338,  0.004648098208404,  0.003162179891248],
+         [-0.997069854672086,  2.006191486385745, -0.990363307725271,  0.011838782789231,  0.009369460553432],
+         [ 0.004226841220338, -0.990363307725271,  2.017072849046023, -0.974622709753106,  0.026274470491696],
+         [ 0.004648098208404,  0.011838782789231, -0.974622709753106,  2.050268428894304, -0.919971837350421],
+         [ 0.003162179891248,  0.009369460553432,  0.026274470491696, -0.919971837350421,  2.269992000979243]]));
+  });
```
Comment (harrysarson) [inline test/function/matrix/sqrtm.test.js]: Can we add a test that sqrtm throws for non square matricies?

File: lib/function/matrix/sqrtm.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+      if (error > _tolerance && ++iterations > _maxIterations) {
+        throw new SyntaxError('Could not converge to solution within the maximum iterations limit');
+      }
+    } while (error > _tolerance);
+
+    return X;
+  }
+
+  sqrtm.toTex = undefined; // use default template
```
Comment (harrysarson) [inline lib/function/matrix/sqrtm.js]: If we are feeling fancy we could probably write the latex of sqrtm as `A^{1/2}`

File: lib/function/matrix/sqrtm.js [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    var X = eye(size(A));
+
+    do {
+      var X_k = X;
+      X = multiply(0.5, add(X, divide(A, X)));
+
+      error = max(abs(subtract(X, X_k)));
+
+      if (error > _tolerance && ++iterations > _maxIterations) {
+        throw new SyntaxError('Could not converge to solution within the maximum iterations limit');
```
Comment (harrysarson) [inline lib/function/matrix/sqrtm.js]: Same as above.

File: lib/function/matrix/sqrtm.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    do {
+      var Y_k = Y;
+      Y = multiply(0.5, add(Y_k, inv(Z)));
+      Z = multiply(0.5, add(Z, inv(Y_k)));
+
+      error = max(abs(subtract(Y, Y_k)));
+
+      if (error > _tolerance && ++iterations > _maxIterations) {
+        throw new SyntaxError('Could not converge to solution within the maximum iterations limit');
```
Comment (harrysarson) [inline lib/function/matrix/sqrtm.js]: This error should be an `Error` and not be a `SyntaxError` and message could be clearer. Maybe: `'Error calculation square root of matrix: iterative method could not converge'` or something similar.

File: lib/function/matrix/sqrtm.js (hunk 1 above)
Comment (ferrolho) [reply lib/function/matrix/sqrtm.js]: This was a copy-paste issue. However, I am not so sure as to which ones to list here... What do you think of listing `sqrt`, `max`, and maybe `abs`?

Comment (harrysarson) [reply lib/function/matrix/sqrtm.js]: Put `pow` in aswell as `math.pow(math.sqrtm(A))` should equal `A` (might be worth testing for this)

**Comment by ferrolho** [comment]
@harrysarson I think all of your suggestions have now been addressed. I do agree they were quite relevant! Thanks! :+1:

**Comment by ferrolho** [comment]
That is correct, `_denmanBeavers` is preferable to `_babylonianMethod` for numerical stability.

I started to implement `_babylonianMethod` before `_denmanBeavers` so I thought *"Why not include both implementations?"*

**Comment by ferrolho** [comment]
Up to you guys. :stuck_out_tongue:

**Comment by josdejong** [comment]
Thanks Henrique, your implementation looks really neat, nice job :+1: 

Thanks @harrysarson for reviewing the PR.

@ferrolho a few more feedback points:

1. > If we are not using _babylonianMethod does it make sense to remove it from the source code? It would make maintaining sqrtm simpler going forward.
    
    Agree, if `_denmanBeavers` is simply the best then let's keep simple and just keep this implementation.

2. Right now you can input a Matrix or an Array, and the output is always a Matrix. I think it would be good to change this behavior such that for Matrix input, the output is a Matrix, and for Array inputs, the output will be an Array, so the behavior is consistent with other functions in math.js.

3. The function is well tested. It would be neat to add one more test: testing a matrix containing BigNumbers instead of regular numbers. Looking at the implementation I think it just works, but it would be good to validate this.

**Comment by ChristopherChudzicki** [comment]
Two pretty minor questions:
1. Right now the tolerance is hard-coded as `1e-6`, right? Is it worth having a configurable tolerance, with `1e-6` as the default?
2. In the `toTex()` method, do you really want the outer parentheses? I'm not sure what's standard typographically, but `math.parse("[[1,2],[3,4]]").toTex()` yields `{\begin{bmatrix}1&2\\3&4\\\end{bmatrix}}^{2}` not `\left( \begin{bmatrix}1&2\\3&4\\\end{bmatrix} \right)^{2}`. (That is, squaring has no outer parentheses.)

**Comment by ferrolho** [comment]
@josdejong I believe the latest commit addressed both points `1` and `3`. Regarding point `2`, I am not so sure if I understood your feedback, as this method is directed towards matrices. For element-wise square root we already have `math.sqrtm(x)`.

I have also edited the LaTeX conversion as pointed out by @ChristopherChudzicki.

**Comment by josdejong** [comment]
Thanks for your updates @ferrolho . I indeed do not mean element-wise computation, sorry for the confusion. The behavior I would expect is basically this:

```js
// Array in -> Array out
var a_in = [[33, 24], [48, 57]]
var a_out = math.sqrtm(a_in)   // outputs a 2D Array [[5,  2], [ 4,  7]]
math.typeof(a_in)              // outputs String 'Array'
math.typeof(a_out)             // outputs String 'Array'

// Matrix in -> Matrix out
var m_in = math.matrix([[33, 24], [48, 57]])
var m_out = math.sqrtm(m_in)   // outputs a 2D Matrix [[5,  2], [ 4,  7]]
math.typeof(m_in)              // outputs String 'Matrix'
math.typeof(m_out)             // outputs String 'Matrix'
```
Does that make sense?

**Comment by ferrolho** [comment]
@josdejong it does make sense. My bad for not understanding it at first...

The last commit should ensure data type persistency. Thanks for your detailed previous comment.

**Comment by josdejong** [comment]
@ferrolho thanks, looks good! that is even less changes in the code needed than I expected :D 

I see one more issue which is important: in the unit tests, at a number of tests `math.deepEqual` is used, which should be `assert.deepEqual`. Currently, these tests do not throw an error at all if the actual result does not match the expected result.

File: test/function/matrix/sqrtm.test.js [hunk 6]
```diff
@@ -11,40 +12,39 @@ describe('sqrtm', function () {
...
+    assert.deepEqual(math.sqrtm([ 4]), [2]);
+    assert.deepEqual(math.sqrtm([16]), [4]);
+    assert.deepEqual(math.sqrtm([20.25]), [4.5]);
   });
 
   it('should return the principal square root of a matrix of big numbers', function() {
-    math.deepEqual(math.sqrtm(math.bignumber(AA)), math.matrix(math.bignumber(A)));
-    math.deepEqual(math.sqrtm(math.bignumber(BB)), math.matrix(math.bignumber(B)));
+    // approx.deepEqual(math.sqrtm(math.bignumber(AA)), math.bignumber(A));
+    // approx.deepEqual(math.sqrtm(math.bignumber(BB)), math.bignumber(B));
```
Comment (ferrolho) [inline test/function/matrix/sqrtm.test.js]: Why is `approx.deepEqual()` not working as expected for *Matrices* of `BigNumbers`?

**Comment by ferrolho** [comment]
@josdejong I believe I have fixed the issue.

However, I do not seem to be able to get the BigNumber unit test to pass — *cf.* the comment above. Am I doing something stupid again? :upside_down_face:

File: test/function/matrix/sqrtm.test.js (hunk 6 above)
Comment (ferrolho) [reply test/function/matrix/sqrtm.test.js]: Hmm... But this goes against the ideology of having input `Array` → `Array` output, and `Matrix` → `Matrix`, no? Both the method I implemented in this PR and the already existing methods should be sound concerning whatever the contents of an `Array` or `Matrix` are, whether they are regular floats or something else (such as a `BigNumber`).

Comment (josdejong) [reply test/function/matrix/sqrtm.test.js]: Ah sorry, I read your changes wrongly, your commented lines should indeed work.

When I do some logging, I see that the output contains round off errors, i.e. for BigNumbers the solution does not fully converge: 

```js
console.log(math.format(math.sqrtm(math.bignumber(AA))))
// [
//   [
//     5.000000000000000000000000931510855380425361089387837696247148415,
//     2.000000000000000000000000931510855380425361089387837696247148416
//   ], 
//   [
//     4.000000000000000000000001863021710760850722178775675392494296833, 
//     7.00000000000000000000000186302171076085072217877567539249429683
//   ]
// ]
```

So we can either find a different example that does converge, or first round the result to say 12 digits using `math.round` and then compare with the expected result.

**Comment by josdejong** [comment]
Thanks @ferrolho, nice job :+1:

**Comment by josdejong** [comment]
@ferrolho `math.sqrtm` is now available in `v4.2.1`

**Pull request merged**